File: 2nd year/MPP/project-mpp/java-project/TriathlonPersistence/src/main/java/triathlon/persistence/dbRepository/DBResultRepository.java
package triathlon.persistence.dbRepository;

import triathlon.model.Athlete;
import triathlon.model.Game;
import triathlon.persistence.validators.ValidationException;
import triathlon.persistence.RepositoryException;
import triathlon.persistence.ResultRepository;
import triathlon.model.Result;
import triathlon.persistence.validators.Validator;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class DBResultRepository implements ResultRepository {
    private final Validator<Result> validator;
    private final Connection connection;

    public DBResultRepository(Properties props, Validator<Result> validator) {
        this.connection = new JDBCManager(props).getConnection();
        this.validator = validator;
    }

    @Override
    public Result findOne(Long id) {
        final String sql_query = "select r.id as id, g.id as gid, a.id as aid, g.name as gname, a.name as aname, value " +
                "from Result r, Game g, Athlete a where id = ? and r.gameId=g.id and r.athleteId=a.id";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);

            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                return new Result(
                        resultSet.getLong("id"),
                        new Game(resultSet.getLong("gid"), resultSet.getString("gname")),
                        new Athlete(resultSet.getLong("aid"), resultSet.getString("aname")),
                        resultSet.getFloat("value")
                );
            }
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }

        return null;
    }

    @Override
    public Iterable<Result> findAll() {
        final String sql_query = "select r.id as id, g.id as gid, a.id as aid, g.name as gname, a.name as aname, value " +
                "from Result r, Game g, Athlete a where r.gameId=g.id and r.athleteId=a.id";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            List<Result> results = new ArrayList<>();
            ResultSet resultSet = statement.executeQuery();
            while (resultSet.next()) {
                results.add(new Result(
                        resultSet.getLong("id"),
                        new Game(resultSet.getLong("gid"), resultSet.getString("gname")),
                        new Athlete(resultSet.getLong("aid"), resultSet.getString("aname")),
                        resultSet.getFloat("value")
                ));
            }
            return results;
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }

        return null;
    }

    @Override
    public void save(Result result) {
        final String sql_query = "insert into Result (gameId, athleteId, value) values (?,?,?)";
        try {
            validator.validate(result);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, result.getGame().getId());
            statement.setLong(2, result.getAthlete().getId());
            statement.setFloat(3, result.getValue());

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void update(Result result) {
        final String sql_query = "update Result set gameId = ?, athleteId = ?, value = ? where id = ?";
        try {
            validator.validate(result);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, result.getGame().getId());
            statement.setLong(2, result.getAthlete().getId());
            statement.setFloat(3, result.getValue());
            statement.setLong(4, result.getId());

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void delete(Long id) {
        final String sql_query = "delete from Result where id = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }
}

File: 2nd year/MPP/project-mpp/csharp-project/persistence/dbRepository/DbResultRepository.cs
using System;
using System.Collections.Generic;
using log4net;
using Microsoft.Data.Sqlite;
using model;
using persistence.validators;

namespace persistence.dbRepository
{
    public class DbResultRepository : IResultRepository
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(DbRefereeRepository));
        private readonly SqliteConnection _connection = JdbcManager.GetConnection();
        private readonly IValidator<Result> _validator;

        public DbResultRepository(IValidator<Result> validator)
        {
            this._validator = validator;
        }

        public Result FindOne(long id)
        {
            Log.Info($"Entered with id = {id}");
            try
            {
                const string commandText = "select * from Result where id = @id";
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                using var reader = command.ExecuteReader();
                if (reader.Read())
                {
                    return new Result
                    {
                        Id = reader.GetInt64(0),
                        Game = new Game
                        {
                            Id = reader.GetInt64(1)
                        },
                        Athlete = new Athlete
                        {
                            Id = reader.GetInt64(2)
                        },
                        Value = reader.GetDouble(3)
                    };
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            return null;
        }

        public IEnumerable<Result> FindAll()
        {
            Log.Info("");
            const string commandText = "select * from Result";
            IList<Result> results = new List<Result>();

            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                using var reader = command.ExecuteReader();
                while (reader.Read())
                {
                    results.Add(new Result
                    {
                        Id = reader.GetInt64(0),
                        Game = new Game
                        {
                            Id = reader.GetInt64(1)
                        },
                        Athlete = new Athlete
                        {
                            Id = reader.GetInt64(2)
                        },
                        Value = reader.GetDouble(3)
                    });
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            Log.Info(results);
            return results;
        }

        public void Save(Result result)
        {
            Log.Info($"Entered with entity = {result}");
            const string commandText =
                "insert into Result (id, gameId, athleteId, value) values (@id, @gameId, @athleteId, @value)";
            _validator.Validate(result);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", result.Id);
                command.Parameters.AddWithValue("gameId", result.Game.Id);
                command.Parameters.AddWithValue("athleteId", result.Athlete.Id);
                command.Parameters.AddWithValue("value", result.Value);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Update(Result result)
        {
            Log.Info($"Entered with entity = {result}");
            const string commandText =
                "update Result set gameId = @gameId, athleteId = @athleteId, value = @value where id = @id";
            _validator.Validate(result);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", result.Id);
                command.Parameters.AddWithValue("gameId", result.Game.Id);
                command.Parameters.AddWithValue("athleteId", result.Athlete.Id);
                command.Parameters.AddWithValue("value", result.Value);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Delete(long id)
        {
            Log.Info($"Entered with id = {id}");
            const string commandText = "delete from Result where id = @id";
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }
    }
}
